### PR TITLE
feat(application): add perception module

### PR DESCRIPTION
**Pull request by kschrab**
## Type of this PR 

- [ ] Bug fix
- [x] Enhancement

## Description

Initial Implementation of a perception module for mosaic-application. 

The perception module models a camera in the vehicle detecting other traffic objects (currently only vehicles) within its field of view. It can be enabled with 

```java
SimplePerceptionConfiguration perceptionModuleConfiguration =
                new SimplePerceptionConfiguration (VIEWING_ANGLE, VIEWING_RANGE);
getOs().getPerceptionModule().enable(perceptionModuleConfiguration);
```

During simulation, the application can query all visible objects with the configured range. Each returned `VehicleObject` provides its position, speed, and heading.

```
List<VehicleObject> perceivedVehicles = getOs().getPerceptionModule().getPerceivedVehicles();
```

The perception module uses the `CentralPerceptionComponent` which uses a spatial index to easily find vehicles within a search range. Current implementations include a `Trivial` approach (for-loop) and a `QuadTree`-based implementation. The implementation can be chosen via `application_config.json`:

```json
{
    "perceptionConfiguration": {
        "perceptionBackend": "QuadTree",
        "performanceMeasure": true,
        "treeSplitSize": 20,
        "treeMaxDepth": 12
    }
}
```

## Issue(s) related to this PR

 * Internal issue 286
 
 
## Affected parts of the online documentation

New documentation has to be added.

## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CameraPerceptionModule.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+        @Override
+        public boolean isInRange(SpatialObject other) {
+            if (other.getId().equals(this.ownerId)) { // cannot see itself
+                return false;
+            }
+            tmpVector2.set(0, 0, 0);
+            synchronized (tmpVector1) {
+                other.getProjectedPosition().toVector3d(tmpVector1).subtract(origin); // convert position of other to relative point
+                return VectorUtils.isLeftOfLine(tmpVector1, tmpVector2, leftBoundVector) // vehicle is left of right edge
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CameraPerceptionModule.java:120]: I renamed leftBoundVector to rightBoundVector and vice-versa.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ErrorRegister.java [hunk 2]
```diff
@@ -45,6 +45,8 @@
     SIMULATION_KERNEL_ConfigurationPathAlreadySet(0x01000027, "The configuration path was already set."),
     SIMULATION_KERNEL_CentralNavigationComponentNotSet(0x01000028, "The CentralNavigationComponent was not set."),
     SIMULATION_KERNEL_CentralNavigationComponentAlreadySet(0x01000029, "The CentralNavigationComponent was already set."),
+    SIMULATION_KERNEL_CentralPerceptionComponentNotSet(0x01000200, "The CentralPerceptionComponent was not set."),
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ErrorRegister.java:48]: It's fine I guess. I don't see any advantage in these error codes anyways.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        if (configuration.measurePerformance) {
+            spatialIndex = new MonitoringSpatialIndex(spatialIndex, performanceMonitor);
+        }
+    }
+
+    public SpatialVehicleIndex getSpatialIndex() {
+        return spatialIndex;
+    }
+
+    public void updateSpatialIndices() {
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/CentralPerceptionComponent.java:105]: I'd keep it for now.

File: test/scenarios/perception-module/application/application_config_grid.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "perceptionConfiguration": {
+        "perceptionBackend": "Grid",
+        "gridCellWidth": "5m",
+        "gridCellHeight": "5m",
+        "spatialIndexUpdateInterval": "1s"
```
Comment (rprotzmann) [inline test/scenarios/perception-module/application/application_config_grid.json:6]: we don't need this parameter configuration anymore

File: test/scenarios/perception-module/application/application_config_quadtree.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+    "perceptionConfiguration": {
+        "perceptionBackend": "QuadTree",
+        "treeSplitSize": 10,
+        "treeMaxDepth": 2,
+        "spatialIndexUpdateInterval": "1s"
```
Comment (rprotzmann) [inline test/scenarios/perception-module/application/application_config_quadtree.json:6]: we don't need this parameter configuration anymore

File: test/scenarios/perception-module/application/application_config_trivial.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
+    "perceptionConfiguration": {
+        "perceptionBackend": "Trivial",
+        "spatialIndexUpdateInterval": "1s"
```
Comment (rprotzmann) [inline test/scenarios/perception-module/application/application_config_trivial.json:4]: we don't need this parameter configuration anymore

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+     */
+    private static class SimplePerception implements PerceptionRange {
+
+        private final String ownerId;
+        private final SimplePerceptionConfiguration configuration;
+
+        private final Vector3d origin = new Vector3d();
+        private final Vector3d rightBoundVector = new Vector3d();
+        private final Vector3d leftBoundVector = new Vector3d();
+        /**
```
Comment (rprotzmann) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/SimplePerceptionModule.java:87]: missing empty line

**Review by rprotzmann — APPROVED** [review]

**Pull request merged**